draft-ietf-quic-ack-frequency.md

```diff
@@ -349,11 +351,17 @@ Unreported Missing:
   have not yet been received.
 
 An endpoint that receives an ACK_FREQUENCY frame with a non-zero Reordering
-Threshold value SHOULD send an immediate ACK when the gap
-between the smallest Unreported Missing packet and the Largest Unacked is greater
-than or equal to the Reordering Threshold value. Sending this additional ACK will
-reset the `max_ack_delay` timer and `Ack-Eliciting Threshold` counter (as any ACK
-would do).
+Threshold value SHOULD send an immediate ACK in response to an Out-of-Order
+Packet:
+
+* when the packet number is larger than the Largest Unacked
+  packet and the difference between the this packet number and the Largest
+  Unacked is larger than the Reordering Threshold.
+
+* when the packet number is lower than the Largest Acked.
+
+* when the total number of Unreported Missing packets is larger
```

Comment:
I think this is redundant to the first bullet if I am reasoning about it correctly?  If there were this many unreported missing packets, wouldn't the first condition be met?

Reply:
Check comment above, the case I am addressing there is:

- threshold is 2.
- the endpoint receives packet 1, no ACK yet because delays. Highest unacked is 1.
- the endpoint receives packet 3, no ACK yet because the gap to 1 is only one packet. Highest unacked is 3.
- the endpoint receives packet 5. There is just a one-packet gap to 3, so according to the first rule, no ACK yet. But there are no two holes, 2 and 4, so the endpoint should send an ACK.

But I agree that's still not right. Consider:

- threshold is 2.
- the endpoint receives packet 1, no ACK yet because delays. Highest unacked is 1.
- the endpoint receives packet 3, no ACK yet because the gap to 1 is only one packet. Highest unacked is 3.
- the endpoint receives packet 4, 5, 6, 7, 8, 9, 10.

They are all in sequence, so the "out of order" rule does not kick in. Also, there is only one hole, so the "number of holes" rule does not kick in either. Yet, if the endpoint acknowledged packet 6, the peer would immediately resend packet 2, because of the number gap rule in RFC 9002. If our goal is to somehow synchronize the sending of ACK with loss detection, then the text ought to focus on the distance between "lowest missing packet number" and "highest unacked".

But then we also need a definition of "lowest missing packet number", and that definition has to consider the "highest number acked". The easiest implementation is to define that as "lowest missing packet number among those larger than largest acknowledged number." That will cover the example above, but there will be corner cases:

- threshold is 2.
- the endpoint receives packet 1, no ACK yet because delays. Highest unacked is 1.
- the endpoint receives packet 3, the endpoint send ACK 1, 3, because of ack delay.
- the endpoint receives packet 4, 5, 6, 7, 8, 9, 10.

Again, if the endpoint sent ACK 1,3-10, then the peer would immediately repeat packet 2 because of the number rule. So maybe the definition should be "lowest missing packet number among those larger than largest acknowledged number minus threshold + 1."

Reply:
The first part is actually fixed in PR #208, which includes the correct test on the "lowest missing number". I will open a new issue for the corner case using the discussion above, and then close this PR.